File: frontend/src/api/monitoring.ts
import { z } from "zod";

import { requestJson } from "./client";

const jsonHeaders = { "Content-Type": "application/json" };

const missionTypeSchema = z.enum([
  "topic",
  "entity",
  "creator",
  "event",
  "research_question",
  "query",
]);
const missionStatusSchema = z.enum([
  "draft",
  "active",
  "paused",
  "running",
  "waiting_platform",
  "waiting_login",
  "completed_run",
  "degraded",
  "failed",
  "archived",
]);
const scheduleSchema = z.enum(["manual", "daily", "weekly", "custom"]);
const budgetSchema = z.object({
  max_model_calls: z.number().int().nonnegative(),
  max_total_tokens: z.number().int().nonnegative(),
  max_collection_count: z.number().int().nonnegative(),
  max_platforms: z.number().int().nonnegative(),
  max_runtime_seconds: z.number().int().positive(),
  daily_token_budget: z.number().int().nonnegative(),
  weekly_run_budget: z.number().int().positive(),
});
const targetSchema = z.object({
  id: z.string().optional(),
  target_type: missionTypeSchema,
  target_value: z.string(),
  normalized_key: z.string().optional(),
  created_at: z.string().optional(),
});

export const monitoringMissionSchema = z.object({
  id: z.string(),
  title: z.string(),
  goal: z.string(),
  mission_type: missionTypeSchema,
  status: missionStatusSchema,
  schedule_type: scheduleSchema,
  schedule_config: z.record(z.string(), z.unknown()),
  platforms: z.array(z.string()),
  understanding: z.record(z.string(), z.unknown()),
  budget: budgetSchema,
  next_run_at: z.string().nullable(),
  last_run_at: z.string().nullable(),
  last_run_status: z.string().nullable(),
  latest_change: z.record(z.string(), z.unknown()).nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const monitoringMissionDetailSchema = monitoringMissionSchema.extend({
  targets: z.array(targetSchema),
  importance_rule: z.string().nullable(),
  ignored_content_rule: z.string().nullable(),
  consecutive_failures: z.number().int().nonnegative(),
  last_error: z.string().nullable(),
});

const runStatusSchema = z.enum([
  "queued",
  "running",
  "waiting_platform",
  "waiting_login",
  "completed",
  "no_meaningful_change",
  "degraded",
  "failed",
  "cancelled",
]);
export const monitoringRunSchema = z.object({
  id: z.string(),
  mission_id: z.string(),
  research_task_id: z.string().nullable(),
  status: runStatusSchema,
  trigger: z.string(),
  started_at: z.string().nullable(),
  completed_at: z.string().nullable(),
  baseline_created: z.boolean(),
  change_count: z.number().int().nonnegative(),
  notification_count: z.number().int().nonnegative(),
  resource: z.record(z.string(), z.unknown()),
  failure_reason: z.string().nullable(),
  backoff_until: z.string().nullable(),
  claimed_at: z.string().nullable(),
  created_at: z.string(),
  queries: z.array(z.record(z.string(), z.unknown())),
});

export const monitoringBaselineSchema = z.object({
  id: z.string(),
  mission_id: z.string(),
  version: z.number().int().positive(),
  snapshot: z.record(z.string(), z.unknown()),
  source_run_id: z.string().nullable(),
  created_at: z.string(),
});

export const monitoringChangeSchema = z.object({
  id: z.string(),
  source_type: z.literal("monitoring"),
  mission_id: z.string(),
  run_id: z.string(),
  change_type: z.string(),
  fingerprint: z.string(),
  title: z.string(),
  summary: z.string(),
  first_seen_at: z.string().nullable(),
  latest_seen_at: z.string().nullable(),
  relevance_score: z.number(),
  novelty_score: z.number(),
  evidence_strength_score: z.number(),
  source_independence_score: z.number(),
  cross_platform_score: z.number(),
  actionability_score: z.number(),
  persistence_score: z.number(),
  noise_risk_score: z.number(),
  attention_level: z.enum([
    "immediate_attention",
    "daily_digest",
    "normal_record",
    "silent_memory",
    "ignored",
  ]),
  state: z.enum(["new", "read", "deferred", "ignored", "merged"]),
  explanation: z.record(z.string(), z.unknown()),
  sources: z.array(z.record(z.string(), z.unknown())),
  memory_update: z.record(z.string(), z.unknown()).nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const monitoringNotificationSchema = z.object({
  id: z.string(),
  mission_id: z.string(),
  change_id: z.string(),
  level: z.enum([
    "immediate_attention",
    "daily_digest",
    "normal_record",
    "silent_memory",
    "ignored",
  ]),
  status: z.enum(["unread", "read", "deferred", "ignored"]),
  title: z.string(),
  summary: z.string(),
  created_at: z.string(),
  read_at: z.string().nullable(),
  deferred_until: z.string().nullable(),
  ignored_at: z.string().nullable(),
});

export const monitoringRunResultSchema = monitoringRunSchema.extend({
  baseline: monitoringBaselineSchema.nullable(),
  changes: z.array(z.record(z.string(), z.unknown())),
  outcome: z.string(),
});

export type MonitoringMission = z.infer<typeof monitoringMissionSchema>;
export type MonitoringMissionDetail = z.infer<typeof monitoringMissionDetailSchema>;
export type MonitoringRun = z.infer<typeof monitoringRunSchema>;
export type MonitoringRunResult = z.infer<typeof monitoringRunResultSchema>;
export type MonitoringBaseline = z.infer<typeof monitoringBaselineSchema>;
export type MonitoringChange = z.infer<typeof monitoringChangeSchema>;
export type MonitoringNotification = z.infer<typeof monitoringNotificationSchema>;

export interface MonitoringMissionInput {
  goal: string;
  title?: string;
  mission_type?: z.infer<typeof missionTypeSchema>;
  targets?: Array<{ target_type: z.infer<typeof missionTypeSchema>; target_value: string }>;
  platforms?: string[];
  schedule_type?: z.infer<typeof scheduleSchema>;
  schedule_config?: Record<string, unknown>;
  importance_rule?: string;
  ignored_content_rule?: string;
  confirmed?: boolean;
}

export function listMonitoringMissions(signal?: AbortSignal) {
  return requestJson("/api/monitoring/missions", z.array(monitoringMissionSchema), { signal });
}

export function createMonitoringMission(input: MonitoringMissionInput) {
  return requestJson("/api/monitoring/missions", monitoringMissionDetailSchema, {
    method: "POST",
    headers: jsonHeaders,
    body: JSON.stringify(input),
  });
}

export function getMonitoringMission(missionId: string, signal?: AbortSignal) {
  return requestJson(
    `/api/monitoring/missions/${encodeURIComponent(missionId)}`,
    monitoringMissionDetailSchema,
    { signal },
  );
}

export function updateMonitoringMission(
  missionId: string,
  input: Record<string, unknown>,
) {
  return requestJson(
    `/api/monitoring/missions/${encodeURIComponent(missionId)}`,
    monitoringMissionDetailSchema,
    { method: "PATCH", headers: jsonHeaders, body: JSON.stringify(input) },
  );
}

export function confirmMonitoringMission(missionId: string) {
  return requestJson(
    `/api/monitoring/missions/${encodeURIComponent(missionId)}/confirm`,
    monitoringMissionDetailSchema,
    { method: "POST", headers: jsonHeaders },
  );
}

export function runMonitoringMission(missionId: string) {
  return requestJson(
    `/api/monitoring/missions/${encodeURIComponent(missionId)}/run`,
    monitoringRunResultSchema,
    { method: "POST", headers: jsonHeaders },
  );
}

function controlMission(missionId: string, action: "pause" | "resume" | "archive") {
  return requestJson(
    `/api/monitoring/missions/${encodeURIComponent(missionId)}/${action}`,
    monitoringMissionDetailSchema,
    { method: "POST", headers: jsonHeaders },
  );
}

export const pauseMonitoringMission = (missionId: string) => controlMission(missionId, "pause");
export const resumeMonitoringMission = (missionId: string) => controlMission(missionId, "resume");
export const archiveMonitoringMission = (missionId: string) => controlMission(missionId, "archive");

export function listMonitoringRuns(missionId: string, signal?: AbortSignal) {
  return requestJson(
    `/api/monitoring/missions/${encodeURIComponent(missionId)}/runs`,
    z.array(monitoringRunSchema),
    { signal },
  );
}

export function listMonitoringChanges(missionId: string, signal?: AbortSignal) {
  return requestJson(
    `/api/monitoring/missions/${encodeURIComponent(missionId)}/changes`,
    z.array(monitoringChangeSchema),
    { signal },
  );
}

export function getMonitoringBaseline(missionId: string, signal?: AbortSignal) {
  return requestJson(
    `/api/monitoring/missions/${encodeURIComponent(missionId)}/baseline`,
    monitoringBaselineSchema.nullable(),
    { signal },
  );
}

export function listMonitoringNotifications(signal?: AbortSignal) {
  return requestJson("/api/notifications", z.array(monitoringNotificationSchema), { signal });
}

export function updateMonitoringNotification(
  notificationId: string,
  action: "read" | "defer" | "ignore",
  until?: string,
) {
  return requestJson(
    `/api/notifications/${encodeURIComponent(notificationId)}/${action}`,
    monitoringNotificationSchema,
    {
      method: "POST",
      headers: jsonHeaders,
      body: JSON.stringify(until ? { until } : {}),
    },
  );
}

File: backend/app/models/monitoring.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

MonitoringMissionType = Literal[
    "topic",
    "entity",
    "creator",
    "event",
    "research_question",
    "query",
]
MonitoringMissionStatus = Literal[
    "draft",
    "active",
    "paused",
    "running",
    "waiting_platform",
    "waiting_login",
    "completed_run",
    "degraded",
    "failed",
    "archived",
]
MonitoringScheduleType = Literal["manual", "daily", "weekly", "custom"]
MonitoringRunStatus = Literal[
    "queued",
    "running",
    "waiting_platform",
    "waiting_login",
    "completed",
    "no_meaningful_change",
    "degraded",
    "failed",
    "cancelled",
]
AttentionLevel = Literal[
    "immediate_attention",
    "daily_digest",
    "normal_record",
    "silent_memory",
    "ignored",
]
NotificationStatus = Literal["unread", "read", "deferred", "ignored"]


class MonitoringTargetInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    target_type: MonitoringMissionType
    target_value: str = Field(min_length=1, max_length=300)

    @field_validator("target_value")
    @classmethod
    def normalize_target(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized or not normalized.isprintable():
            raise ValueError("target_value must be printable and non-blank")
        return normalized


class MonitoringBudget(BaseModel):
    model_config = ConfigDict(extra="forbid")

    max_model_calls: int = Field(default=4, ge=0, le=20)
    max_total_tokens: int = Field(default=8_000, ge=0, le=100_000)
    max_collection_count: int = Field(default=3, ge=0, le=20)
    max_platforms: int = Field(default=3, ge=0, le=7)
    max_runtime_seconds: int = Field(default=300, ge=1, le=3_600)
    daily_token_budget: int = Field(default=20_000, ge=0, le=1_000_000)
    weekly_run_budget: int = Field(default=7, ge=1, le=31)


class MonitoringMissionCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    goal: str = Field(min_length=5, max_length=10_000)
    title: str | None = Field(default=None, max_length=200)
    mission_type: MonitoringMissionType = "research_question"
    targets: list[MonitoringTargetInput] = Field(default_factory=list, max_length=8)
    platforms: list[str] = Field(default_factory=list, max_length=7)
    schedule_type: MonitoringScheduleType = "manual"
    schedule_config: dict[str, object] = Field(default_factory=dict)
    importance_rule: str | None = Field(default=None, max_length=1_000)
    ignored_content_rule: str | None = Field(default=None, max_length=1_000)
    budget: MonitoringBudget = Field(default_factory=MonitoringBudget)
    confirmed: bool = False

    @field_validator("goal")
    @classmethod
    def normalize_goal(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized or not normalized.isprintable():
            raise ValueError("goal must be printable and non-blank")
        return normalized

    @field_validator("platforms")
    @classmethod
    def normalize_platforms(cls, values: list[str]) -> list[str]:
        normalized = [value.strip().casefold() for value in values]
        if any(not value or not value.isprintable() for value in normalized):
            raise ValueError("platforms must be printable and non-blank")
        if len(set(normalized)) != len(normalized):
            raise ValueError("platforms must be unique")
        return normalized


class MonitoringMissionPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, max_length=200)
    schedule_type: MonitoringScheduleType | None = None
    schedule_config: dict[str, object] | None = None
    platforms: list[str] | None = Field(default=None, max_length=7)
    importance_rule: str | None = Field(default=None, max_length=1_000)
    ignored_content_rule: str | None = Field(default=None, max_length=1_000)
    budget: MonitoringBudget | None = None


class MonitoringMissionSummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    title: str
    goal: str
    mission_type: MonitoringMissionType
    status: MonitoringMissionStatus
    schedule_type: MonitoringScheduleType
    schedule_config: dict[str, object]
    platforms: list[str]
    understanding: dict[str, object]
    budget: MonitoringBudget
    next_run_at: str | None
    last_run_at: str | None
    last_run_status: str | None
    latest_change: dict[str, object] | None
    created_at: str
    updated_at: str


class MonitoringMissionDetail(MonitoringMissionSummary):
    targets: list[dict[str, object]]
    importance_rule: str | None
    ignored_content_rule: str | None
    consecutive_failures: int
    last_error: str | None


class MonitoringRun(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    mission_id: str
    research_task_id: str | None
    status: MonitoringRunStatus
    trigger: str
    started_at: str | None
    completed_at: str | None
    baseline_created: bool
    change_count: int
    notification_count: int
    resource: dict[str, object]
    failure_reason: str | None
    backoff_until: str | None
    claimed_at: str | None
    created_at: str
    queries: list[dict[str, object]]


class MonitoringChange(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    source_type: Literal["monitoring"] = "monitoring"
    mission_id: str
    run_id: str
    change_type: str
    fingerprint: str
    title: str
    summary: str
    first_seen_at: str | None
    latest_seen_at: str | None
    relevance_score: float
    novelty_score: float
    evidence_strength_score: float
    source_independence_score: float
    cross_platform_score: float
    actionability_score: float
    persistence_score: float
    noise_risk_score: float
    attention_level: AttentionLevel
    state: Literal["new", "read", "deferred", "ignored", "merged"]
    explanation: dict[str, object]
    sources: list[dict[str, object]]
    memory_update: dict[str, object] | None
    created_at: str
    updated_at: str


class MonitoringNotification(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    mission_id: str
    change_id: str
    level: AttentionLevel
    status: NotificationStatus
    title: str
    summary: str
    created_at: str
    read_at: str | None
    deferred_until: str | None
    ignored_at: str | None


class MonitoringBaseline(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    mission_id: str
    version: int
    snapshot: dict[str, object]
    source_run_id: str | None
    created_at: str


class MonitoringRunResult(MonitoringRun):
    baseline: MonitoringBaseline | None = None
    changes: list[dict[str, object]] = Field(default_factory=list)
    outcome: str


class NotificationAction(BaseModel):
    model_config = ConfigDict(extra="forbid")

    until: str | None = Field(default=None, max_length=80)
